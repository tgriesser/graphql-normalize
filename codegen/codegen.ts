import type { CodegenPlugin } from '@graphql-codegen/plugin-helpers';

const NormalizeCodegenPlugin: CodegenPlugin = {
  plugin(schema, documents, config, info) {
    //
  },
};

export default NormalizeCodegenPlugin;
