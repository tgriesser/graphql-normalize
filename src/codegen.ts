import { NormalizeCodegenPlugin } from './codegen/plugin.js';
export type { TypePolicies } from './codegen/getCacheKey.js';
export type { NormalizeMetaShape, NormalizedDocShape } from './metadataShapes.js';

export { NormalizeCodegenPlugin } from './codegen/plugin.js';
export { generateNormalizedMetadataForDocs, generateNormalizedMetadata } from './codegen/generateNormalizedMetadata.js';
export { generateNormalizedOperations, generateNormalizedOperation } from './codegen/generateNormalizedOperation.js';

export { NormalizeCodegenPlugin as plugin };

export default {
  plugin: NormalizeCodegenPlugin,
};
