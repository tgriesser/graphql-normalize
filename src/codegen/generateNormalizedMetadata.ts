import { optimizeDocuments } from '@graphql-tools/relay-operation-optimizer'
import {
  DocumentNode,
  GraphQLSchema,
  Kind,
  TypeInfo,
  isListType,
  isWrappingType,
  visit,
  print,
  visitWithTypeInfo,
  ArgumentNode,
  FieldNode,
  OperationTypeNode,
  isObjectType,
  isScalarType,
  InlineFragmentNode,
  getNamedType,
  assertObjectType,
  isEnumType,
} from 'graphql'
import type {
  FieldDef,
  FieldMeta,
  NormalizeMetaShape,
  NormalizedDocShape,
  UnionMeta,
  VariableMeta,
} from '../metadataShapes'
import { TypePolicies, getCacheKey } from './getCacheKey'
import { stringifyVariables } from '../stringifyVariables'

export function generateNormalizedMetadata(
  schema: GraphQLSchema,
  document: DocumentNode,
  typePolicies: TypePolicies = {}
) {
  return generateNormalizedMetadataForDocs(schema, [document], typePolicies)[0].meta
}

/**
 * Given an operation and a schema, generates the metadata necessary to
 * write the operation into the cache
 */
export function generateNormalizedMetadataForDocs(
  schema: GraphQLSchema,
  documents: DocumentNode[],
  typePolicies: TypePolicies = {}
): NormalizedDocShape[] {
  let finalDocuments = documents
  if (documents.some((o) => o.definitions.some((d) => d.kind === Kind.FRAGMENT_DEFINITION))) {
    finalDocuments = optimizeDocuments(schema, documents, {
      noLocation: true,
      includeFragments: false,
    })
  }

  const normalizedDocs: NormalizedDocShape[] = []

  const typeInfo = new TypeInfo(schema)
  let normalizedDoc: NormalizeMetaShape = {
    operation: OperationTypeNode.QUERY,
    variables: [],
    fields: [],
  }

  let parentStack: Array<FieldMeta | UnionMeta | NormalizeMetaShape> = []
  let parentFieldDef: FieldMeta | NormalizeMetaShape | UnionMeta = normalizedDoc

  function popStack() {
    const parent = parentStack.pop()
    if (parent) {
      parentFieldDef = parent
    }
  }

  function pushField(field: string | FieldDef) {
    parentFieldDef.fields ??= []
    if (typeof field === 'string' && parentFieldDef.fields.includes(field)) {
      return
    }
    parentFieldDef.fields.push(field)
    if (typeof field !== 'string') {
      parentStack.push(parentFieldDef)
      parentFieldDef = field
    }
  }

  function normalizeArgs(nodes: readonly ArgumentNode[]) {
    let hasVar = false
    const argsDef: Record<string, any> = {}
    for (const arg of nodes) {
      if (arg.value.kind === Kind.VARIABLE) {
        argsDef[`$${arg.value.name.value}`] = arg.name.value
        hasVar = true
      } else {
        argsDef[arg.name.value] = print(arg.value)
      }
    }
    return hasVar ? argsDef : stringifyVariables(argsDef)
  }

  const visitor = visitWithTypeInfo(typeInfo, {
    OperationDefinition: {
      enter(op) {
        normalizedDoc.operation = op.operation
      },
      leave(op) {
        normalizedDocs.push({
          name: op.name?.value ?? 'unnamed',
          meta: normalizedDoc,
        })
        normalizedDoc = {
          operation: OperationTypeNode.QUERY,
          variables: [],
          fields: [],
        }
        parentFieldDef = normalizedDoc
        parentStack = []
      },
    },
    VariableDefinition(node) {
      const obj: VariableMeta = {
        name: node.variable.name.value,
      }
      if (node.defaultValue) {
        obj.defaultValue = print(node.defaultValue)
      }
      normalizedDoc.variables.push(obj)
    },
    Field: {
      enter(node) {
        const { gqlType, listDepth } = unpackType(typeInfo)

        // If this is a boring ol scalar / enum, we push it as a string
        if (isSimpleField(node, typeInfo)) {
          pushField(node.name.value)
          return
        }

        const fieldMeta: FieldMeta = { name: node.name.value }
        pushField(fieldMeta)

        if (listDepth) fieldMeta.list = listDepth
        if (node.alias?.value) fieldMeta.alias = node.alias.value

        if (isObjectType(gqlType)) {
          const cacheKey = getCacheKey(typePolicies, gqlType)
          if (cacheKey) {
            fieldMeta.cacheKey = cacheKey
          }
          if (gqlType.name === schema.getQueryType()?.name) {
            fieldMeta.cacheKey = '$root'
          }
        }
        if (node.arguments?.length) {
          fieldMeta.args = normalizeArgs(node.arguments)
        }

        const skipDirective = node.directives?.find((d) => d.name.value === 'skip')
        const includeDirective = node.directives?.find((d) => d.name.value === 'include')
        if (skipDirective) {
          fieldMeta.skip = normalizeArgs(skipDirective.arguments ?? [])
        }
        if (includeDirective) {
          fieldMeta.include = normalizeArgs(includeDirective.arguments ?? [])
        }
      },
      leave(node) {
        // Pop things back into place
        if (!isSimpleField(node, typeInfo)) {
          popStack()
        }
      },
    },
    InlineFragment: {
      enter(node) {
        if (
          node.typeCondition?.name.value &&
          isConcreteAbstract(node, typeInfo, schema) &&
          parentIsFieldMeta(parentFieldDef)
        ) {
          const unionType: UnionMeta = {
            fields: [],
          }
          const outputType = getNamedType(typeInfo.getType())
          const cacheKey = getCacheKey(typePolicies, assertObjectType(outputType))
          if (cacheKey) {
            unionType.cacheKey = cacheKey
          }

          parentFieldDef.possible ??= {}
          parentFieldDef.possible[node.typeCondition?.name.value] = unionType
          parentStack.push(parentFieldDef)
          parentFieldDef = unionType
        }
      },
      leave(node) {
        if (node.typeCondition?.name.value && isConcreteAbstract(node, typeInfo, schema)) {
          popStack()
        }
      },
    },
    FragmentDefinition() {
      throw noFragments()
    },
  })

  for (const doc of finalDocuments) {
    visit(doc, visitor)
  }

  return normalizedDocs
}

function unpackType(typeInfo: TypeInfo) {
  let listDepth = 0
  // We want to unwrap any list types, and make those the containing
  let gqlType = typeInfo.getType()
  while (isWrappingType(gqlType)) {
    if (isListType(gqlType)) {
      ++listDepth
    }
    gqlType = gqlType.ofType
  }
  return { gqlType, listDepth }
}

function isSimpleField(node: FieldNode, typeInfo: TypeInfo) {
  const { gqlType, listDepth } = unpackType(typeInfo)
  return (
    (isScalarType(gqlType) || isEnumType(gqlType)) &&
    !listDepth &&
    !node.alias &&
    !node.arguments?.length &&
    !node.directives?.some((d) => d.name.value === 'skip' || d.name.value === 'includes')
  )
}

function noFragments() {
  return new Error('Fragment definitions have been stripped')
}

function isConcreteAbstract(node: InlineFragmentNode, typeInfo: TypeInfo, schema: GraphQLSchema) {
  if (node.typeCondition?.name) {
    const parentTypeName = typeInfo.getParentType()?.name
    if (node.typeCondition.name.value !== parentTypeName) {
      return isObjectType(schema.getType(node.typeCondition?.name.value))
    }
  }
  return false
}

function parentIsFieldMeta(parent: FieldMeta | NormalizeMetaShape | UnionMeta): parent is FieldMeta {
  return 'name' in parent
}
