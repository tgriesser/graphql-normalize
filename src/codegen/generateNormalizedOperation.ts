import { optimizeDocuments } from '@graphql-tools/relay-operation-optimizer';
import { DocumentNode, GraphQLSchema, Kind } from 'graphql';
import { addNormalizingFields } from './addNormalizingFields';
import type { TypePolicies } from './getCacheKey';

export function generateNormalizedOperation(
  schema: GraphQLSchema,
  document: DocumentNode,
  typePolicies: TypePolicies = {}
) {
  return generateNormalizedOperations(schema, [document], typePolicies)[0];
}

/**
 * Given a schema & operation, flattens the fragments and adds any necessary missing fields
 * in order to properly normalize the response in a cache.
 */
export function generateNormalizedOperations(
  schema: GraphQLSchema,
  documents: DocumentNode[],
  typePolicies: TypePolicies = {}
) {
  let toNormalize = documents;
  if (documents.some((o) => o.definitions.some((d) => d.kind === Kind.FRAGMENT_DEFINITION))) {
    toNormalize = optimizeDocuments(schema, documents, {
      noLocation: true,
      includeFragments: false,
    });
  }
  return addNormalizingFields(schema, toNormalize, typePolicies);
}
