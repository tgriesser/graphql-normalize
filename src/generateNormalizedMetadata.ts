import { optimizeDocuments } from '@graphql-tools/relay-operation-optimizer';
import {
  DocumentNode,
  GraphQLSchema,
  Kind,
  TypeInfo,
  isAbstractType,
  isListType,
  isWrappingType,
  visit,
  print,
  visitWithTypeInfo,
  ArgumentNode,
  FieldNode,
  SelectionSetNode,
  OperationTypeNode,
  isObjectType,
} from 'graphql';
import stableStringify from 'fast-json-stable-stringify';
import type { FieldDef, FieldMeta, NormalizeMeta, NormalizedDoc, VariableMeta } from './extensionShape';
import { TypePolicies, getCacheKey } from './typePolicies';

/**
 * Given an operation and a schema, generates the metadata necessary to
 * write the operation into the cache
 */
export function generateNormalizedMetadata(
  schema: GraphQLSchema,
  operation: DocumentNode,
  typePolicies: TypePolicies = {}
): NormalizedDoc {
  let finalOperation = operation;
  if (operation.definitions.some((d) => d.kind === Kind.FRAGMENT_DEFINITION)) {
    finalOperation = optimizeDocuments(schema, [operation], {
      noLocation: true,
    })[0];
  }

  const typeInfo = new TypeInfo(schema);
  const normalizedDoc: NormalizedDoc = {
    operation: OperationTypeNode.QUERY,
    variables: [],
    selectionSet: {},
  };

  const fieldPath: any[] = [];

  function setPath<K extends keyof NormalizeMeta>(key: K, value: NormalizeMeta[K]) {
    normalizedDoc.selectionSet[fieldPath.join('.')] ??= {};
    normalizedDoc.selectionSet[fieldPath.join('.')][key] = value;
  }
  function normalizeArgs(nodes: readonly ArgumentNode[]) {
    let hasVar = false;
    const argsDef: Record<string, any> = {};
    for (const arg of nodes) {
      if (arg.value.kind === Kind.VARIABLE) {
        argsDef[`$${arg.name.value}`] = arg.name.value;
        hasVar = true;
      } else {
        argsDef[arg.name.value] = print(arg.value);
      }
    }
    return hasVar ? argsDef : stableStringify(argsDef);
  }
  function normalizeField(node: FieldNode) {
    if (node.alias || node.arguments?.length) {
      const field: FieldMeta = {
        name: node.name.value,
      };
      if (node.alias) field.alias = node.alias.value;
      if (node.arguments?.length) field.args = normalizeArgs(node.arguments);
      return field;
    }
    return node.name.value;
  }
  function addSelectionFields(node: SelectionSetNode) {
    const fields: FieldDef[] = [];
    for (const sel of node.selections) {
      if (sel.kind === Kind.FRAGMENT_SPREAD) {
        throw noFragments();
      }
      if (sel.kind === Kind.FIELD) {
        fields.push(normalizeField(sel));
      }
      // If we have an inline fragment
      if (sel.kind === Kind.INLINE_FRAGMENT) {
        //
      }
    }
    setPath('fields', fields);
  }

  const visitor = visitWithTypeInfo(typeInfo, {
    OperationDefinition(op) {
      normalizedDoc.operation = op.operation;
    },
    VariableDefinition(node) {
      const obj: VariableMeta = {
        name: node.variable.name.value,
      };
      if (node.defaultValue) {
        obj.defaultValue = print(node.defaultValue);
      }
      normalizedDoc.variables.push(obj);
    },
    Field: {
      enter(node) {
        if (node.alias?.value) {
          fieldPath.push(node.alias.value);
          setPath('aliasField', node.name.value);
        } else {
          fieldPath.push(node.name.value);
        }
        let gqlType = typeInfo.getType();
        while (isWrappingType(gqlType)) {
          if (isListType(gqlType)) {
            setPath('list', true);
            fieldPath.push('$idx');
          }
          gqlType = gqlType.ofType;
        }
        if (isAbstractType(gqlType)) {
          setPath(
            'possible',
            schema.getPossibleTypes(gqlType).map((t) => t.name)
          );
        }
        if (isObjectType(gqlType)) {
          const cacheKey = getCacheKey(typePolicies, gqlType);
          if (cacheKey) {
            setPath('cacheKey', cacheKey);
          }
        }
        if (node.arguments?.length) {
          setPath('args', normalizeArgs(node.arguments));
        }
        if (node.selectionSet) {
          addSelectionFields(node.selectionSet);
        }
      },
      leave(node) {
        fieldPath.pop();
        let gqlType = typeInfo.getType();
        while (isWrappingType(gqlType)) {
          if (isListType(gqlType)) {
            fieldPath.pop();
          }
          gqlType = gqlType.ofType;
        }
      },
    },
    InlineFragment: {
      enter(node) {
        if (node.typeCondition?.name) {
          fieldPath.push(`$type_${node.typeCondition?.name.value}`);
          addSelectionFields(node.selectionSet);
        }
      },
      leave(node) {
        if (node.typeCondition?.name.value) {
          fieldPath.pop();
        }
      },
    },
    FragmentDefinition() {
      throw noFragments();
    },
  });

  visit(finalOperation, visitor);

  return normalizedDoc;
}

function noFragments() {
  return new Error('Fragment definitions have been stripped');
}
