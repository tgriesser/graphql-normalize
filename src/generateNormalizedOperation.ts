import { optimizeDocuments } from '@graphql-tools/relay-operation-optimizer';
import { DocumentNode, GraphQLSchema, Kind } from 'graphql';
import { addNormalizingFields } from './addNormalizingFields';

export function generateNormalizedOperation(schema: GraphQLSchema, operation: DocumentNode) {
  let finalOperation = operation;
  if (operation.definitions.some((d) => d.kind === Kind.FRAGMENT_DEFINITION)) {
    finalOperation = optimizeDocuments(schema, [operation], {
      noLocation: true,
    })[0];
  }
  return addNormalizingFields(schema, finalOperation);
}
