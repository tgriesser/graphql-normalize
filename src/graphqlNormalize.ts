import type { FieldDef, FieldMeta, NormalizeMetaShape } from './metadataShapes';
import type { FormattedExecutionResult } from 'graphql';
import type { CacheShape } from './cache';
import { __typename } from './constants';
import { getArgsObj, printArgs } from './printArgs';
import { makeCacheKey } from './makeCacheKey';

type Path = Array<string | number>;

export interface SyncWithCacheOptions {
  // read = cache overwrites result
  // write = result overwrites cache
  action: 'read' | 'write';
  // Used to determine the args, directives, etc.
  variableValues: Record<string, any>;
  // Shape of the metadata for the current operation
  // we're reading or writing
  meta: NormalizeMetaShape;
  // The shape of the normalized cache
  cache: CacheShape;
  // The result we're writing into the cache, if any
  operationResult?: FormattedExecutionResult;
  // The current result, used as the target object to mutate,
  // or a new value if we're initially syncing
  currentResult?: FormattedExecutionResult['data'];
  // Equality function, used to deal with scalars
  isEqual?: (a: any, b: any) => boolean;
}

interface BaseConfig {
  cacheVal: any;
  resultVal: any;
  targetVal: any;
}

interface HandleFieldConfig extends BaseConfig {
  field: FieldMeta;
}

interface WriteListFieldConfig extends HandleFieldConfig {
  atIndex: number;
}

interface SyncTraverseOptions extends BaseConfig {
  fields: FieldDef[];
}

interface TraverseListConfig {
  field: FieldMeta;
  cacheVal: Array<any>;
  targetVal: Array<any>;
  resultVal: Array<any>;
  depth: number;
  currentDepth?: number;
}

function defaultIsEqual(a: any, b: any) {
  return a === b;
}

// Ensure the path exists in an object
const ensure = (obj: any, path: Path, fallback: any) => {
  let source = obj;
  for (let i = 0; i < path.length; i++) {
    const key = path[i];
    const hasNext = i < path.length - 1;
    if (hasNext) {
      if (source[key] === undefined) {
        source[key] = typeof key === 'number' ? [] : {};
      }
      source = source[key];
    } else if (source[key] === undefined) {
      source[key] = fallback;
      return fallback;
    } else {
      return source[key];
    }
  }
};

const shouldSkipField = (field: FieldMeta, meta: NormalizeMetaShape, variableValues: any) => {
  if (!field.skip && !field.include) {
    return false;
  }
  const argsObj = getArgsObj(field.skip ?? field.include, meta, variableValues);
  if (field.skip) {
    return argsObj.if;
  }
  return !argsObj.if;
};

const get = (obj: any, key: string | number) => {
  return obj[key];
};

const getIn = (obj: any, path: Path) => {
  let source = obj;
  for (let i = 0; i < path.length; i++) {
    const key = path[i];
    if (source[key] == null) {
      return source[key];
    }
    source = source[key];
  }
  return source;
};

export interface SyncWithCacheResult {
  added: number;
  updated: number;
  cache: CacheShape;
  result: Exclude<FormattedExecutionResult['data'], null>;
}

export function graphqlNormalize(options: SyncWithCacheOptions): SyncWithCacheResult {
  const {
    action,
    variableValues,
    cache,
    operationResult = { data: {}, errors: [] },
    currentResult = {},
    isEqual = defaultIsEqual,
    meta,
  } = options;

  let added = 0;
  let updated = 0;
  const isWrite = action === 'write';
  const isRead = !isWrite;

  const set = (obj: any, key: string | number, val: any) => {
    if (!isEqual(obj[key], val)) {
      obj[key] = val;
    }
  };

  const setIn = (obj: any, path: Path, value: any) => {
    let source = obj;
    for (let i = 0; i < path.length; i++) {
      const key = path[i];
      const hasNext = i < path.length - 1;
      if (hasNext) {
        if (source[key] === undefined) {
          source[key] = typeof key === 'number' ? [] : {};
        }
        source = source[key];
      } else if (!isEqual(source[key], value)) {
        if (source[key] === undefined) {
          added++;
        } else {
          updated++;
        }
        source[key] = value;
      }
    }
  };

  const traverseList = (config: TraverseListConfig) => {
    const { field, resultVal, targetVal, cacheVal, depth, currentDepth = 1 } = config;
    if (resultVal.length < targetVal.length) {
      targetVal.length = resultVal.length;
    }
    if (isWrite && resultVal.length < cacheVal.length) {
      cacheVal.length = resultVal.length;
    }

    for (let i = 0; i < resultVal.length; i++) {
      if (currentDepth < depth && resultVal[i] !== null) {
        if (isWrite) {
          ensure(cacheVal, [i], []);
        }
        ensure(targetVal, [i], []);
        traverseList({
          field,
          depth,
          resultVal: resultVal[i],
          targetVal: targetVal[i],
          cacheVal: cacheVal[i],
          currentDepth: currentDepth + 1,
        });
      } else {
        writeListField({
          field,
          cacheVal,
          targetVal,
          resultVal,
          atIndex: i,
        });
      }
    }
  };

  const writeListField = (writeConfig: WriteListFieldConfig) => {
    const { cacheVal, resultVal, targetVal, atIndex, field } = writeConfig;

    const { cacheKeyVal, fields, typename } = isRead
      ? readFieldMeta(field, resultVal[atIndex])
      : writeFieldMeta(field, resultVal[atIndex]);
    if (isWrite) {
      if (cacheKeyVal) {
        set(cacheVal, atIndex, { $ref: cacheKeyVal });
        setIn(cache.fields, [cacheKeyVal, __typename], typename);
      } else {
        set(cacheVal, atIndex, resultVal[atIndex]);
      }
    }

    if (fields) {
      ensure(targetVal, [atIndex], {});
      traverseFields({
        fields,
        targetVal: ensure(targetVal, [atIndex], {}),
        resultVal: resultVal[atIndex],
        cacheVal: cacheKeyVal ? ensure(cache.fields, [cacheKeyVal], {}) : ensure(cacheVal, [atIndex], {}),
      });
    } else {
      set(targetVal, atIndex, resultVal[atIndex]);
    }
  };

  const readFieldMeta = (field: FieldMeta, val: any) => {
    const { possible, cacheKey, fields } = field;
    if (!cacheKey && !val?.$ref) {
      return {
        typename: undefined,
        fields,
        cacheKeyVal: undefined,
      };
    }
    const typename = val.$ref?.split(':')[0] ?? val[__typename];
    if (possible && val.$ref) {
      const possibleInfo = possible[typename];
      if (possibleInfo) {
        return {
          typename,
          fields: possibleInfo.fields.concat(fields ?? []),
          cacheKeyVal: val.$ref,
        };
      }
    }
    return {
      typename,
      fields,
      cacheKeyVal: val.$ref,
    };
  };

  const writeFieldMeta = (field: FieldMeta, val: any) => {
    const { possible, cacheKey, fields } = field;
    const typename = val?.[__typename];
    if (possible && val) {
      const possibleInfo = possible[typename];
      if (possibleInfo) {
        return {
          typename,
          fields: possibleInfo.fields.concat(fields ?? []),
          cacheKeyVal: possibleInfo.cacheKey ? makeCacheKey(possibleInfo.cacheKey, val) : undefined,
        };
      }
    }
    return {
      typename,
      fields: fields,
      cacheKeyVal: cacheKey ? makeCacheKey(cacheKey, val) : undefined,
    };
  };

  const handleTraverseField = (field: FieldMeta, options: SyncTraverseOptions) => {
    const { cacheVal, resultVal, targetVal } = options;
    const resultName = field.alias ?? field.name;
    const argsKey = printArgs(field.args, meta, variableValues);

    const handleField = (writeConfig: HandleFieldConfig) => {
      const { cacheVal, resultVal, targetVal, field } = writeConfig;
      const { cacheKeyVal, fields, typename } = isRead
        ? readFieldMeta(field, getIn(cacheVal, [field.name, argsKey]))
        : writeFieldMeta(field, resultVal[resultName]);

      if (isWrite) {
        if (cacheKeyVal) {
          setIn(cacheVal, [field.name, argsKey], { $ref: cacheKeyVal });
          setIn(cache.fields, [cacheKeyVal, __typename], typename);
        }
      }

      if (fields) {
        const cacheFieldVal = cacheKeyVal
          ? getIn(cache.fields, [cacheKeyVal])
          : ensure(cacheVal, [field.name, argsKey], {});
        traverseFields({
          fields: fields,
          cacheVal: cacheFieldVal,
          targetVal: ensure(targetVal, [resultName], {}),
          resultVal: isRead ? cacheFieldVal : ensure(resultVal, [resultName], {}),
        });
      } else {
        if (isWrite) {
          setIn(cacheVal, [field.name, argsKey], resultVal[resultName]);
        }
        set(targetVal, resultName, getIn(cacheVal, [field.name, argsKey]));
      }
    };

    if (field.list) {
      // If we're writing, we're setting the cache, so we use the cache as the
      // target val while also setting the value in result
      if (isWrite) {
        ensure(cacheVal, [field.name, argsKey], []);
      }
      ensure(targetVal, [resultName], []);

      // We iterate the list, copying the old values to the new values
      traverseList({
        field,
        depth: field.list,
        resultVal: isRead ? getIn(cacheVal, [field.name, argsKey]) : ensure(resultVal, [resultName], []),
        targetVal: targetVal[resultName],
        cacheVal: ensure(cacheVal, [field.name, argsKey], []),
      });
    } else {
      handleField({
        field,
        resultVal,
        targetVal,
        cacheVal,
      });
    }
  };

  const traverseFields = (options: SyncTraverseOptions) => {
    const { fields, cacheVal, resultVal, targetVal } = options;
    for (const field of fields) {
      if (field === __typename) {
        if (isWrite) {
          set(cacheVal, __typename, get(resultVal, __typename));
        }
        set(targetVal, __typename, get(cacheVal, __typename));
      } else if (typeof field === 'string') {
        if (isWrite) {
          setIn(cacheVal, [field, '$'], getIn(resultVal, [field]));
        }
        set(targetVal, field, getIn(cacheVal, [field, '$']));
      } else if (shouldSkipField(field, meta, variableValues)) {
        delete targetVal[field.alias ?? field.name];
      } else {
        handleTraverseField(field, options);
      }
    }
  };

  traverseFields({
    fields: meta.fields,
    cacheVal: cache.fields,
    resultVal: operationResult.data,
    targetVal: currentResult,
  });

  return { added, updated, cache, result: currentResult ?? {} };
}
