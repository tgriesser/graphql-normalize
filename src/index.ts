import { graphqlNormalize } from './graphqlNormalize.js';
export { stringifyVariables } from './stringifyVariables.js';
export { MissingFieldError } from './errors.js';
export type { graphqlNormalizeOptions, NormalizeMetaShape } from './graphqlNormalize.js';
export { graphqlNormalize };
export default graphqlNormalize;
