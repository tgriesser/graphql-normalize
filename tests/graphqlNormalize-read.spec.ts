import { execute } from 'graphql';
import { beforeEach, describe, expect, it } from 'vitest';
import { produce } from 'immer';

import { generateNormalizedOperation } from '../src/codegen/generateNormalizedOperation';
import { schema } from './fixtures/schema';
import { operation1Doc } from './fixtures/ops';
import { generateNormalizedMetadata } from '../src/codegen/generateNormalizedMetadata';
import type { NormalizeMetaShape } from '../src/metadataShapes';
import { graphqlNormalize } from '../src/graphqlNormalize';

describe('graphqlNormalize: read', () => {
  let cache = {
    operations: {},
    fields: {},
  };
  let meta: NormalizeMetaShape;
  beforeEach(async () => {
    meta = generateNormalizedMetadata(schema, operation1Doc);
    const variableValues = {};
    cache = {
      operations: {},
      fields: {},
    };
    graphqlNormalize({
      action: 'write',
      meta,
      operationResult: await execute({
        schema,
        variableValues,
        document: generateNormalizedOperation(schema, operation1Doc),
      }),
      variableValues,
      cache,
    });
  });

  it('builds a cache of values', async () => {
    const { result } = graphqlNormalize({
      action: 'read',
      meta: produce(meta, () => {}), // ensures meta isn't mutated
      cache: produce(cache, () => {}), // ensures cache isn't mutated
      variableValues: {},
    });
    expect(result).toMatchSnapshot();
  });

  it('does not mutate the data if it has not changed', async () => {
    const { result: readResult } = graphqlNormalize({
      action: 'read',
      meta,
      cache,
      variableValues: {},
    });
    const initial = produce({}, () => readResult);
    const secondRead = produce(readResult, (existing) => {
      graphqlNormalize({
        action: 'read',
        meta,
        cache,
        currentResult: existing,
        variableValues: {},
      });
    });
    expect(secondRead).toEqual(initial);
  });
});
