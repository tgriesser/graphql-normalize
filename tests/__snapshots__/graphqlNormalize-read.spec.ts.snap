// Vitest Snapshot v1

exports[`graphqlNormalize: read > builds a cache of values 1`] = `
{
  "enum": "a",
  "first10Posts": {
    "nodes": [
      {
        "id": "UG9zdDox",
        "title": "Some Blog post 1",
      },
      {
        "id": "UG9zdDoy",
        "title": "Some Blog post 2",
      },
      {
        "id": "UG9zdDoz",
        "title": "Some Blog post 3",
      },
      {
        "id": "UG9zdDo0",
        "title": "Some Blog post 4",
      },
      {
        "id": "UG9zdDo1",
        "title": "Some Blog post 5",
      },
      {
        "id": "UG9zdDo2",
        "title": "Some Blog post 6",
      },
      {
        "id": "UG9zdDo3",
        "title": "Some Blog post 7",
      },
      {
        "id": "UG9zdDo4",
        "title": "Some Blog post 8",
      },
      {
        "id": "UG9zdDo5",
        "title": "Some Blog post 9",
      },
      {
        "id": "UG9zdDoxMA==",
        "title": "Some Blog post 10",
      },
    ],
  },
  "homeItems": {
    "nodes": [
      {
        "__typename": "Comment",
      },
      {
        "id": "QmxvZzox",
        "name": "My Cool #1",
      },
      {
        "__typename": "Post",
        "id": "UG9zdDox",
      },
      {
        "__typename": "User",
      },
      {
        "__typename": "Comment",
      },
      {
        "id": "QmxvZzoy",
        "name": "My Cool #2",
      },
      {
        "__typename": "Post",
        "id": "UG9zdDoy",
      },
      {
        "__typename": "User",
      },
      {
        "__typename": "Comment",
      },
      {
        "id": "QmxvZzoz",
        "name": "My Cool #3",
      },
    ],
  },
  "nestedString": [
    [
      "a",
      "b",
    ],
    [
      "foo",
    ],
  ],
  "nestedUserList": [
    [
      {
        "name": "user 1",
      },
    ],
    [
      {
        "name": "user 2",
      },
    ],
  ],
  "num": 1,
  "posts": {
    "edges": [
      {
        "cursor": "T2Zmc2V0Q29ubmVjdGlvbjow",
        "node": {
          "author": {
            "id": "VXNlcjox",
          },
          "id": "UG9zdDox",
        },
      },
      {
        "cursor": "T2Zmc2V0Q29ubmVjdGlvbjox",
        "node": {
          "author": {
            "id": "VXNlcjoy",
          },
          "id": "UG9zdDoy",
        },
      },
      {
        "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoy",
        "node": {
          "author": {
            "id": "VXNlcjox",
          },
          "id": "UG9zdDoz",
        },
      },
      {
        "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoz",
        "node": {
          "author": {
            "id": "VXNlcjoy",
          },
          "id": "UG9zdDo0",
        },
      },
      {
        "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo0",
        "node": {
          "author": {
            "id": "VXNlcjox",
          },
          "id": "UG9zdDo1",
        },
      },
      {
        "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo1",
        "node": {
          "author": {
            "id": "VXNlcjoy",
          },
          "id": "UG9zdDo2",
        },
      },
      {
        "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo2",
        "node": {
          "author": {
            "id": "VXNlcjox",
          },
          "id": "UG9zdDo3",
        },
      },
      {
        "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo3",
        "node": {
          "author": {
            "id": "VXNlcjoy",
          },
          "id": "UG9zdDo4",
        },
      },
      {
        "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo4",
        "node": {
          "author": {
            "id": "VXNlcjox",
          },
          "id": "UG9zdDo5",
        },
      },
      {
        "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo5",
        "node": {
          "author": {
            "id": "VXNlcjoy",
          },
          "id": "UG9zdDoxMA==",
        },
      },
      {
        "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoxMA==",
        "node": {
          "author": {
            "id": "VXNlcjox",
          },
          "id": "UG9zdDoxMQ==",
        },
      },
      {
        "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoxMQ==",
        "node": {
          "author": {
            "id": "VXNlcjoy",
          },
          "id": "UG9zdDoxMg==",
        },
      },
    ],
    "nodes": [
      {
        "author": {
          "id": "VXNlcjox",
        },
        "id": "UG9zdDox",
      },
      {
        "author": {
          "id": "VXNlcjoy",
        },
        "id": "UG9zdDoy",
      },
      {
        "author": {
          "id": "VXNlcjox",
        },
        "id": "UG9zdDoz",
      },
      {
        "author": {
          "id": "VXNlcjoy",
        },
        "id": "UG9zdDo0",
      },
      {
        "author": {
          "id": "VXNlcjox",
        },
        "id": "UG9zdDo1",
      },
      {
        "author": {
          "id": "VXNlcjoy",
        },
        "id": "UG9zdDo2",
      },
      {
        "author": {
          "id": "VXNlcjox",
        },
        "id": "UG9zdDo3",
      },
      {
        "author": {
          "id": "VXNlcjoy",
        },
        "id": "UG9zdDo4",
      },
      {
        "author": {
          "id": "VXNlcjox",
        },
        "id": "UG9zdDo5",
      },
      {
        "author": {
          "id": "VXNlcjoy",
        },
        "id": "UG9zdDoxMA==",
      },
      {
        "author": {
          "id": "VXNlcjox",
        },
        "id": "UG9zdDoxMQ==",
      },
      {
        "author": {
          "id": "VXNlcjoy",
        },
        "id": "UG9zdDoxMg==",
      },
    ],
  },
}
`;
